# 跟數學有關的模組
## 數學模組-math
python內建的工具模組之一(e.g.數學-math、統計-statistics、時間-time .etc)，如果想用要引入：

`import math`

### 四捨五入 & 無條件進位 & 無條件捨去
#### 四捨五入-round()
語法：`round(number, digits)`

digits 四捨五入取到小數點後第幾位

#### 無條件進位 & 無條件捨去-math.ceil() & math.floor()
1. 引入math：`import math`
2. 只能取到整數，如要取到小數點後要自己寫

##### 無條件進位

語法：`math.ceil(num)`

##### 無條件捨去

語法：`math.floor(num)`

In [1]:
import math

#四捨五入
a=1.1564846456
print(round(a,2))

#無條件進位
a=1.1564846456
print(math.ceil(a*100)/100)

#無條件捨去
a=1.1564846456
print(math.floor(a*100)/100)

1.16
1.16
1.15


### 圓周率 & log
`math.pi`：圓周率

`math.e`：log

In [2]:
print(math.pi)
print(math.e)

3.141592653589793
2.718281828459045


### 弧度&角度
`math.radians(num)`：角度(num)轉弧度

`math.degrees(num)`：弧度(num)轉角度

In [3]:
print(math.radians(30))
print(math.degrees(0.5235987755982988))

0.5235987755982988
29.999999999999996


### sin & cos & tan
x=弧度

`math.sin(x)`：sin(x)的角度

`math.cos(x)`：cos(x)的角度

`math.tan(x)`：tan(x)的角度

### 反的sin & cos & tan

`math.asin(x)`：用sin(x)的角度求sin(x)的弧度

`math.acos(x)`：用cos(x)的角度求cos(x)的弧度

`math.atan(x)`：用tan(x)的角度求tan(x)的弧度

In [4]:
#假設一個30度的直角三角形邊分別是1:math.sqrt(3):2
a=math.radians(30)
print(a)
print(f"{math.sin(a)} -> 1/2\n{math.cos(a)} -> math.sqrt(3)/2\n{math.tan(a)} -> math.sqrt1/(3)")
print(math.asin(0.49999999999999994))

0.5235987755982988
0.49999999999999994 -> 1/2
0.8660254037844387 -> math.sqrt(3)/2
0.5773502691896257 -> math.sqrt1/(3)
0.5235987755982988


## 統計模組-statistics
算鐘形曲線的，更專業要參考Numpy、SciPy。
### 平均值 & 中位數 & 眾數
`statistics.mean([數列])`：平均值

`statistics.median([數列])`：中位數

`statistics.mode([數列])`：眾數，數列中重複出現最多的數字

**偶個數時(數列有偶數個e.g.10個)**

`statistics.median_high([數列])`：較高的中位數

`statistics.median_low([數列])`：較低的中位數

In [5]:
import statistics

print(f"平均值：{statistics.mean([1,2,3,5,7,11,13,17,19])}") #平均值

a=[1,3,5,7,9,2,4,6,8,10]
print(f"中位數：{statistics.median(a)}") #中位數
print(f"較高的中位數：{statistics.median_high(a)}") #較高的中位數
print(f"較低的中位數：{statistics.median_low(a)}") #較低的中位數

print(f"眾數：{statistics.mode([1,2,2,3,3,3,4,4,4,5,6,7,8,5,5,5,5,6,8,8,7,7,7,7,1])}") 

平均值：8.666666666666666
中位數：5.5
較高的中位數：6
較低的中位數：5
眾數：5


### 母體標準差、變異數
`statistics.pstdev([數列])`：母體標準差=標準差

概念：全校500人，500人全部身高的標準差

定義：跟標準(平均)差多少 -> [離均差的平方/n]開根號=(平方合的平均-平均的平方)開根號

`statistics.pvariance([數列])`：母體變異數 [離均差的平方/n]

範例：

[1,3,4,4,3,9]

平均=(1+3+4+4+3+9)/6=4

標準差=`{[(1-4)**2+(3-4)**2+(4-4)**2+(4-4)**2+(3-4)**2+(9-4)**2]/6}**1/2=6**1/2`

變異數=6

### 樣本標準差、變異數
`statistics.stdev([數列])`：樣本標準差 跟樣本標準(平均)差多少 -> [離均差的平方/n-1]開根號

概念：全校500人，抽50人出來身高的標準差

`statistics.variance([數列])`：樣本變異數 [離均差的平方/n-1]

範例：

[1,3,4,4,3,9]

平均=(1+3+4+4+3+9)/6=4

標準差=`{[(1-4)**2+(3-4)**2+(4-4)**2+(4-4)**2+(3-4)**2+(9-4)**2]/5}**1/2=(36/5)**1/2`

變異數=36/5

至於為何樣本標準差是n-1，可以看這個：[從標準差除以n或除以n-1談起 丁春成](https://web.math.sinica.edu.tw/math_media/d291/29102.pdf)

In [6]:
arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]
a = statistics.pstdev(arr)
b = statistics.pvariance(arr)
c = statistics.stdev(arr)
d = statistics.variance(arr)
print(a)
print(b)
print(c)
print(d)

2.581988897471611
6.666666666666667
2.7386127875258306
7.5


## 精準度模組-decimal
decimal分成2個模塊：

Decimal：精準計算

getcontext：建構環境

引入要特別寫：

In [7]:
from decimal import Decimal, getcontext

### decimal的算術語境
也就是getcontext()裡可以修改的地方

In [8]:
from decimal import *
getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

`prec=28` 精準度(precision)，預設28位。

`rounding=ROUND_HALF_EVEN` 四捨五入的方式：偶數捨入法。

1. 如果要捨入的數字（小數點後的位數）為5，則向最接近的偶數捨入。如果要捨入的數字為奇數，則向下捨入到最接近的偶數；如果要捨入的數字為偶數，則向上捨入到最接近的偶數。

2. 對於非5的數字，按照傳統的四捨五入規則進行捨入。

    範例：

    4.5捨入為4：因為5是奇數，所以向下捨入。

    5.5捨入為6：因為5是奇數，所以向上捨入。

    8.4捨入為8：按照傳統四捨五入規則，捨入到最接近的整數，因為4小於5。

`Emin=-999999, Emax=999999` 最小指數：-999999~最大指數：999999。

`capitals=1` 科學記號表示法用大寫E (e.g.1E+3) ;capitals=0表示用小寫e (e.g.1e+3)。

`clamp=0` 不截斷計算成果。

`flags=[]` 這是一個列表，用於存儲標誌或標記，通常用於標識特定的計算條件或狀態。但在這沒有提供具體的使用情境。

`traps=[InvalidOperation, DivisionByZero, Overflow]` 

可能發生的異常狀況：無效操作、除以0、溢出，會觸發上面這三個異常處理程序。

### 有個小限制
float轉換成二進位儲存時，會有點變化

In [9]:
a = 1.115
print(f"a：{Decimal(a)}") # Decimal(a) 的值會比較奇異一點

b = 1.125
print(f"b：{Decimal(b)}") # Decimal(b) 的值就沒變

a：1.1149999999999999911182158029987476766109466552734375
b：1.125


**解決方法：傳入str**

In [10]:
a = 1.115
print(f"a：{Decimal(str(a))}")

b = 1.125
print(f"b：{Decimal(str(b))}")

a：1.115
b：1.125


### 小數點後的精準四捨五入-quantize
#### round的限制：二進位儲存造成結果偏誤
round約整的方法並不是四捨五入，是偶數捨入法，也就是decimal的ROUND_HALF_EVEN，這個約整方法在round無法變更。

In [11]:
print(f"a：{round(1.115, 2)}")
# 這是因為在電腦內部，1.115 真正的值是 1.1149999999999999911182158029987476766109466552734375

print(f"b：{round(1.125, 2)}")
# 1.125 真正的值也是 1.125 

a：1.11
b：1.12


**解決方法：Decimal.quantize用str輸入** + ROUND_HALF_UP

總覺得有點不實際

In [12]:
print(f"a：{Decimal('1.115').quantize(Decimal('0.00'), ROUND_HALF_UP)}")

print(f"b：{Decimal('1.125').quantize(Decimal('0.00'), ROUND_HALF_EVEN)}")

a：1.12
b：1.12


## 亂數-random
各種隨機

引入：`import random`
### 初始化隨機生成器-seed
1. seed = 亂數的演算法。通常以電腦系統的時間為主，這個可以改。
2. seed相同，產生的答案都一樣。

In [8]:
import random
random.seed(1)
print(random.random())
random.seed(1)
print(random.random())

0.13436424411240122
0.13436424411240122


### 隨機數-random
隨機產生0.0~1.0之間隨機浮點數。

In [9]:
print(random.random())
print(random.random())

0.8474337369372327
0.763774618976614


### 兩數之間的隨機**浮點數**-uniform
語法：`random.uniform(x,y)`，會產生x~y之間的隨機浮點數。

In [12]:
print(random.uniform(3,5))
print(random.uniform(3,5))

3.9908701741838817
3.898982129577476


### 兩數之間的隨機**整數**-randint
語法：`random.randint(x,y)`，會產生x~y之間的隨機整數。

In [13]:
print(random.randint(3,10))
print(random.randint(3,10))

9
6


### 進階-兩數之間的隨機**整數**-randrange
語法1：`random.range(x,y,間隔多少數字)`，會間隔產生x~y之間的隨機整數。

語法：`random.range(y)`，會產生`0`~y之間的隨機整數。

In [15]:
print(random.randrange(1,99,2))
print(random.randrange(1,99,2))
print(random.randrange(13))

49
55
9


### 隨機抽取串列中項目-choice
從str、tuple、list(有順序的物件)中隨機抽取項目。

dict要寫list & keys/values

In [26]:
print(random.choice("123456"))
print(random.choice((1,2,3,4,5,6)))
print(random.choice([1,2,3,4,5,6]))

aaa={"a":1,"b":2,"c":3,"d":4,"e":5}
print(random.choice(list(aaa.keys()))) # 印key
print(random.choice(list(aaa.values()))) # 印value

5
2
3
b
2


### 進階-隨機抽取k個串列中項目-choices
1. 從str、tuple、list(有順序的物件)中隨機抽取k個項目。
    1. **k可能是重複值**。
    2. k會組成一個新的串列。
2. 語法：random.choice(seq,k=要取出多少個項目)
3. 一樣，dict要寫list & keys/values

In [28]:
print(random.choices("123456",k=3))
print(random.choices((1,2,3,4,5,6),k=3))
print(random.choices([1,2,3,4,5,6],k=3))

aaa={"a":1,"b":2,"c":3,"d":4,"e":5}
print(random.choices(list(aaa.keys()),k=3)) # 印key
print(random.choices(list(aaa.values()),k=3)) # 印value

['5', '6', '6']
[3, 6, 6]
[1, 4, 5]
['b', 'd', 'e']
[5, 3, 5]


### choice不會重複的朋友-sample
1. 從str、tuple、list(有順序的物件)中隨機抽取k個項目。
    1. **k不會重複**。
    2. k會組成一個新的串列。
    3. **k一定要填**，不能大於串列長度。
2. 語法：random.choice(seq,k=要取出多少個項目)
3. 一樣，dict要寫list & keys/values

In [29]:
print(random.sample("123456",k=3))
print(random.sample((1,2,3,4,5,6),k=3))
print(random.sample([1,2,3,4,5,6],k=3))

aaa={"a":1,"b":2,"c":3,"d":4,"e":5}
print(random.sample(list(aaa.keys()),k=3)) # 印key
print(random.sample(list(aaa.values()),k=3)) # 印value

['5', '2', '3']
[3, 5, 4]
[5, 4, 1]
['d', 'b', 'c']
[4, 5, 3]


將原本串列打亂-shuffle

# 從這裡開始還沒寫完
# sequence